utils/src/main/scala/za/co/absa/enceladus/utils/time/DateTimePattern.scala

```diff
@@ -17,72 +17,140 @@ package za.co.absa.enceladus.utils.time
+    override val epochFactor: Long = 0
+    override val epochMilliFactor: Long = 0
+
+    override val timeZoneInPattern: Boolean = DateTimePattern.timeZoneInPattern(pattern)
+    override val defaultTimeZone: Option[String] = assignedDefaultTimeZone.filterNot(_ => timeZoneInPattern)
+    override val isTimeZoned: Boolean = timeZoneInPattern || defaultTimeZone.nonEmpty
+  }
+
+  private def createDateTimePattern(pattern: String,
+                                    assignedDefaultTimeZone: Option[String],
+                                    isDefault: Boolean): DateTimePattern = {
+    if (isEpoch(pattern)) {
+      EpochDTPattern(pattern, isDefault)
+    } else {
+      StandardDTPattern(pattern, assignedDefaultTimeZone, isDefault)
+    }
   }
 
-  def isEpoch(format: String): Boolean = {
-    format.toLowerCase match {
-      case "epoch" | "epochmilli" => true
+  def apply(pattern: String,
+            assignedDefaultTimeZone: Option[String] = None): DateTimePattern = {
+    createDateTimePattern(pattern, assignedDefaultTimeZone, isDefault = false)
+  }
+
+  def fromStructField(structField: StructField ): DateTimePattern = {
+    if (!Seq(DateType, TimestampType).contains(structField.dataType)) {
+      val typeName = structField.dataType.typeName
+      throw new InvalidParameterException(
+        s"StrucField data type for DateTimePattern has to be DateType or TimestampType, instead $typeName was given."
+      )
+    }
+    val PatternInfo(pattern, isDefault) = structField.readPatternInfo()
+    val timeZoneOpt = structField.getMetadata("timezone")
+    createDateTimePattern(pattern, timeZoneOpt, isDefault)
+  }
+
+  def isEpoch(pattern: String): Boolean = {
+    pattern.toLowerCase match {
+      case EpochKeyword | EpochMilliKeyword => true
       case _ => false
     }
   }
 
-  def epochFactor(format: String): Long = {
-    format.toLowerCase match {
-      case "epoch"      => epochUnitFactor
-      case "epochmilli" => epochThousandFactor
-      case _            => throw new InvalidParameterException(s"'$format' is not an epoch pattern")
+  def epochFactor(pattern: String): Long = {
+    pattern.toLowerCase match {
+      case EpochKeyword      => epochUnitFactor
+      case EpochMilliKeyword => epochThousandFactor
+      case _                 => 0
     }
   }
 
-  def epochMilliFactor(format: String): Long = {
-    format.toLowerCase match {
-      case "epoch"      => epochThousandFactor
-      case "epochmilli" => epochUnitFactor
-      case _            => throw new InvalidParameterException(s"'$format' is not an epoch pattern")
+  def epochMilliFactor(pattern: String): Long = {
+    pattern.toLowerCase match {
+      case EpochKeyword      => epochThousandFactor
+      case EpochMilliKeyword => epochUnitFactor
+      case _                 => 0
     }
   }
+
+  def timeZoneInPattern(pattern: String): Boolean = {
+    if (isEpoch(pattern)) {
+      true
+    } else {
+      pattern.hasUnquoted(patternTimeZoneChars, Set('''))
```

Comment:
This could be just a oneliner `isEpoch(pattern) || pattern.hasUnquoted(patternTimeZoneChars, Set('''))`. I for some reason don't like the if X then true style.

Reply:
It's much nicer indeed. But it either relies on `||` optimization when in standard form or it always runs into the `pattern.hasUnquoted` evaluation. So wondering, what is the less bad...

Reply:
Both Java and Scala have `||` optimized and I would not worry about it being changed. It is a feature of the languages that gets a lot of praise.

Reply:
This `|` is an eager OR (or a bitwise operator), which evaluates both sides of the operator, `||` is a short-circuit OR. No need to worry in C-derived languages.

Reply:
To by technically pedantic the `||` function operator can be overloaded, and then it stops to be smart enough not to evaluate the right side unless needed.
But on the other hand I think it's safe to assume that for `Boolean` it's not probable, and if done, the author deserves the consequences. Will change it, you were right.